File: legacy/v0.0.8/ppf-contact-solver/eigsys/eigsys_3.cpp
// File: eigsys_3.cpp
// Code: Claude Code and Codex
// Review: Ryoichi Ando (ryoichi.ando@zozo.com)
// License: Apache v2.0

#include "eig-hpp/eigsolve3x3.hpp"
#include <Eigen/Dense>
#include <array>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace Eigen;
static double h = 1e-4;
inline double sqr(double x) { return x * x; }
using Vector9d = Vector<double, 9>;
using Matrix9d = Matrix<double, 9, 9>;

double energy_s(double a, double b, double c, const std::string &m) {
    if (m == "ARAP") {
        return sqr(a - 1) + sqr(b - 1) + sqr(c - 1);
    } else if (m == "SymDirichlet") {
        return sqr(a) + 1 / sqr(a) + sqr(b) + 1 / sqr(b) + sqr(c) + 1 / sqr(c);
    } else if (m == "MIPS") {
        return (sqr(a) + sqr(b) + sqr(c)) / (a * b * c);
    } else if (m == "Ogden") {
        double sum = 0.0;
        for (int k = 0; k < 5; ++k) {
            sum += std::pow(a, std::pow(0.5, k)) +
                   std::pow(b, std::pow(0.5, k)) +
                   std::pow(c, std::pow(0.5, k)) - 3;
        }
        return sum;
    } else if (m == "Yeoh") {
        double sum = 0.0;
        for (int k = 0; k < 3; ++k) {
            sum += std::pow(sqr(a) + sqr(b) + sqr(c) - 3, k + 1);
        }
        return sum;
    } else {
        return 0.0;
    }
}

double energy_F(const Matrix3d &F, const std::string &m) {
    auto [_U, lambda, _Vt] = run_svd_3x3(F);
    return energy_s(lambda[0], lambda[1], lambda[2], m);
}

Vector3d approx_grad_s(double a, double b, double c, const std::string &m) {
    return (Vector3d() << (energy_s(a + h, b, c, m) -
                           energy_s(a - h, b, c, m)) /
                              (2 * h),
            (energy_s(a, b + h, c, m) - energy_s(a, b - h, c, m)) / (2 * h),
            (energy_s(a, b, c + h, m) - energy_s(a, b, c - h, m)) / (2 * h))
        .finished();
}

Matrix3d approx_hess_s(double a, double b, double c, const std::string &m) {
    Matrix3d H;
    H.col(0) = (approx_grad_s(a + h, b, c, m) - approx_grad_s(a - h, b, c, m)) /
               (2 * h);
    H.col(1) = (approx_grad_s(a, b + h, c, m) - approx_grad_s(a, b - h, c, m)) /
               (2 * h);
    H.col(2) = (approx_grad_s(a, b, c + h, m) - approx_grad_s(a, b, c - h, m)) /
               (2 * h);
    return H;
}

double approx_grad_F(const Matrix3d &F, const Matrix3d &dF,
                     const std::string &m) {
    return (energy_F(F + h * dF, m) - energy_F(F - h * dF, m)) / (2 * h);
}

Matrix9d approx_hess_F(const Matrix3d &F, const std::array<Matrix3d, 9> &dF,
                       const std::string &m) {
    int len = dF.size();
    Matrix9d H = Matrix9d::Zero(len, len);
    for (int i = 0; i < len; ++i) {
        for (int j = 0; j < len; ++j) {
            H(i, j) = (approx_grad_F(F + h * dF[i], dF[j], m) -
                       approx_grad_F(F - h * dF[i], dF[j], m)) /
                      (2 * h);
        }
    }
    return H;
}

Matrix3d gen_dF(int i, int j) {
    Matrix3d dF = Matrix3d::Zero(3, 3);
    dF(i, j) = 1;
    return dF;
}

Vector9d mat2vec(const Matrix3d &A) {
    Vector9d x;
    int index = 0;
    for (int j = 0; j < A.cols(); ++j) {
        for (int i = 0; i < A.rows(); ++i) {
            x(index++) = A(i, j);
        }
    }
    return x;
}

int main() {
    std::srand(std::time(0));
    std::cout << std::scientific;
    std::cout << std::setprecision(3);

    Matrix3d F = Matrix3d::Random(3, 3);
    bool verbose = false;
    std::array<std::string, 5> models = {"ARAP", "SymDirichlet", "MIPS",
                                         "Ogden", "Yeoh"};

    std::array<Matrix3d, 9> dF;
    for (int j = 0; j < 3; ++j) {
        for (int i = 0; i < 3; ++i) {
            dF[i + 3 * j] = (gen_dF(i, j));
        }
    }

    auto [U, sigma, Vt] = run_svd_3x3(F);
    std::map<std::string, Vector2d> errors;
    for (const auto &m : models) {
        Vector9d g_F = Vector9d::Zero();
        for (int i = 0; i < 9; ++i) {
            g_F[i] = approx_grad_F(F, dF[i], m);
        }
        if (verbose) {
            std::cout << "---- (" << m << ") numerical gradient ----\n";
            std::cout << g_F << std::endl;
        }

        Matrix9d H_F = approx_hess_F(F, dF, m);
        if (verbose) {
            std::cout << "---- (" << m << ") numerical hessian ----\n";
            std::cout << H_F << std::endl;
        }

        Matrix3d H_s = approx_hess_s(sigma[0], sigma[1], sigma[2], m);
        Vector3d g_s = approx_grad_s(sigma[0], sigma[1], sigma[2], m);

        auto [S_s, U_s] = sym_eigsolve_3x3(H_s);
        Vector9d g_F_rebuilt = Vector9d::Zero();
        for (int i = 0; i < 9; ++i) {
            for (int k = 0; k < 3; ++k) {
                g_F_rebuilt[i] +=
                    g_s[k] * U.col(k).dot(dF[i] * Vt.row(k).transpose());
            }
        }
        if (verbose) {
            std::cout << "--- (" << m << ") analytical gradient ---\n";
            std::cout << g_F_rebuilt << std::endl;
        }

        double inv_sqrt2 = 1.0 / std::sqrt(2);
        std::array<Matrix3d, 9> Q = {
            {inv_sqrt2 * U *
                 (Matrix3d() << 0, 1, 0, -1, 0, 0, 0, 0, 0).finished() * Vt,
             inv_sqrt2 * U *
                 (Matrix3d() << 0, 0, 1, 0, 0, 0, -1, 0, 0).finished() * Vt,
             inv_sqrt2 * U *
                 (Matrix3d() << 0, 0, 0, 0, 0, 1, 0, -1, 0).finished() * Vt,
             inv_sqrt2 * U *
                 (Matrix3d() << 0, 1, 0, 1, 0, 0, 0, 0, 0).finished() * Vt,
             inv_sqrt2 * U *
                 (Matrix3d() << 0, 0, 1, 0, 0, 0, 1, 0, 0).finished() * Vt,
             inv_sqrt2 * U *
                 (Matrix3d() << 0, 0, 0, 0, 0, 1, 0, 1, 0).finished() * Vt,
             U * U_s.col(0).asDiagonal() * Vt, U * U_s.col(1).asDiagonal() * Vt,
             U * U_s.col(2).asDiagonal() * Vt}};

        Vector<double, 9> lmds(9);
        lmds << (g_s[0] + g_s[1]) / (sigma[0] + sigma[1]),
            (g_s[0] + g_s[2]) / (sigma[0] + sigma[2]),
            (g_s[1] + g_s[2]) / (sigma[1] + sigma[2]),
            (std::abs(sigma[0] - sigma[1]) > h
                 ? (g_s[0] - g_s[1]) / (sigma[0] - sigma[1])
                 : H_s(0, 0) - H_s(0, 1)),
            (std::abs(sigma[0] - sigma[2]) > h
                 ? (g_s[0] - g_s[2]) / (sigma[0] - sigma[2])
                 : H_s(0, 0) - H_s(0, 2)),
            (std::abs(sigma[1] - sigma[2]) > h
                 ? (g_s[1] - g_s[2]) / (sigma[1] - sigma[2])
                 : H_s(1, 1) - H_s(1, 2)),
            S_s[0], S_s[1], S_s[2];

        Matrix9d H_rebuilt = Matrix9d::Zero();
        for (int i = 0; i < 9; ++i) {
            if (lmds[i]) {
                Map<Vector9d> q(Q[i].data());
                H_rebuilt += lmds[i] * q * q.transpose();
            }
        }

        if (verbose) {
            std::cout << "--- (" << m << ") analytical hessian ---\n";
            std::cout << H_rebuilt << std::endl;
        }

        errors[m] = Vector2d((g_F - g_F_rebuilt).norm() / g_F.norm(),
                             (H_F - H_rebuilt).norm() / H_F.norm());
    }

    std::cout << "===== error summary =====\n";
    for (const auto &[name, err] : errors) {
        std::cout << name << ": " << std::endl << err << std::endl;
    }

    return 0;
}

File: legacy/v0.0.8/ppf-contact-solver/eigsys/eigsys_3.py
# File: eigsys_3.py
# Code: Claude Code and Codex
# Review: Ryoichi Ando (ryoichi.ando@zozo.com)
# License: Apache v2.0

import numpy as np
import numpy.linalg as LA
from eig_py.eigsolve3x3 import sym_eigsolve_3x3

h = 1e-4
F = np.random.rand(3, 3)
verbose = False
models = [
    "ARAP",
    "SymDirichlet",
    "MIPS",
    "Ogden",
    "Yeoh",
]

# ==== uncommnent below to try a case two singular values are almost the same =====
# a, b, c = 0.73000001, 0.73, 0.780002
# U, _, Vt = LA.svd(F)
# F = U @ np.diag([a, b, c]) @ Vt


def svd3x3(F):
    A = F.T @ F
    lmd, eigvecs = sym_eigsolve_3x3(A)
    sigma = np.sqrt(lmd)
    U = F @ eigvecs
    for i in range(3):
        U[:, i] /= np.linalg.norm(U[:, i])
    return U, sigma, eigvecs.T


def energy_s(a, b, c, m):
    if m == "ARAP":
        return sum([(s - 1) ** 2 for s in [a, b, c]])
    elif m == "SymDirichlet":
        return sum([s**2 + 1 / s**2 for s in [a, b, c]])
    elif m == "MIPS":
        return sum([s**2 for s in [a, b, c]]) / (a * b * c)
    elif m == "Ogden":
        return sum(
            [a ** (0.5**k) + b ** (0.5**k) + c ** (0.5**k) - 3 for k in range(5)]
        )
    elif m == "Yeoh":
        return sum([(a**2 + b**2 + c**2 - 3) ** (k + 1) for k in range(3)])
    else:
        return 0.0


def energy_F(F, m):
    _, (a, b, c), _ = svd3x3(F)
    return energy_s(a, b, c, m)


def approx_grad_s(a, b, c, m):
    return np.array(
        [
            (energy_s(a + h, b, c, m) - energy_s(a - h, b, c, m)) / (2 * h),
            (energy_s(a, b + h, c, m) - energy_s(a, b - h, c, m)) / (2 * h),
            (energy_s(a, b, c + h, m) - energy_s(a, b, c - h, m)) / (2 * h),
        ]
    )


def approx_hess_s(a, b, c, m):
    H = np.zeros((3, 3))
    H[:, 0] = (approx_grad_s(a + h, b, c, m) - approx_grad_s(a - h, b, c, m)) / (2 * h)
    H[:, 1] = (approx_grad_s(a, b + h, c, m) - approx_grad_s(a, b - h, c, m)) / (2 * h)
    H[:, 2] = (approx_grad_s(a, b, c + h, m) - approx_grad_s(a, b, c - h, m)) / (2 * h)
    return H


def approx_grad_F(F, dF, m):
    return (energy_F(F + h * dF, m) - energy_F(F - h * dF, m)) / (2 * h)


def approx_hess_F(F, dF, m):
    H = np.zeros((len(dF), len(dF)))
    for i, dFi in enumerate(dF):
        for j, dFj in enumerate(dF):
            H[i][j] = (
                approx_grad_F(F + h * dFi, dFj, m) - approx_grad_F(F - h * dFi, dFj, m)
            ) / (2 * h)
    return H


def gen_dF(i, j):
    dF = np.zeros((3, 3))
    dF[i][j] = 1
    return dF


def mat2vec(A):
    x = []
    for j in range(3):
        for i in range(3):
            x.append(A[i][j])
    return np.array(x)


dF = []
for j in range(3):
    for i in range(3):
        dF.append(gen_dF(i, j))

U, (a, b, c), Vt = svd3x3(F)
errors = {}
for m in models:
    g_F = np.zeros(9)
    for i, dFi in enumerate(dF):
        g_F[i] = approx_grad_F(F, dFi, m)
    g_F = np.reshape(g_F, (3, 3)).T
    if verbose:
        print(f"---- ({m}) numerical gradient ----")
        print(g_F)

    H_F = approx_hess_F(F, dF, m)
    if verbose:
        print(f"---- ({m}) numerical hessian ----")
        print(H_F)

    ####### Analytical Eigen Decomposition #######
    H_s, g_s = approx_hess_s(a, b, c, m), approx_grad_s(a, b, c, m)
    S_s, U_s = sym_eigsolve_3x3(H_s)

    g_F_rebuilt = np.zeros(9)
    for i, dFi in enumerate(dF):
        g_F_rebuilt[i] = sum(
            g_s[k] * np.dot(U[:, k], np.dot(dFi, Vt[k, :])) for k in range(3)
        )
    g_F_rebuilt = np.reshape(g_F_rebuilt, (3, 3)).T
    if verbose:
        print(f"---- ({m}) analytical gradient ----")
        print(g_F_rebuilt)

    Qs = [
        np.array([[0, 1, 0], [-1, 0, 0], [0, 0, 0]]) / np.sqrt(2),
        np.array([[0, 0, 1], [0, 0, 0], [-1, 0, 0]]) / np.sqrt(2),
        np.array([[0, 0, 0], [0, 0, 1], [0, -1, 0]]) / np.sqrt(2),
        np.array([[0, 1, 0], [1, 0, 0], [0, 0, 0]]) / np.sqrt(2),
        np.array([[0, 0, 1], [0, 0, 0], [1, 0, 0]]) / np.sqrt(2),
        np.array([[0, 0, 0], [0, 0, 1], [0, 1, 0]]) / np.sqrt(2),
        np.diag(U_s[:, 0]),
        np.diag(U_s[:, 1]),
        np.diag(U_s[:, 2]),
    ]
    lmds = [
        (g_s[0] + g_s[1]) / (a + b),
        (g_s[0] + g_s[2]) / (a + c),
        (g_s[1] + g_s[2]) / (b + c),
        (g_s[0] - g_s[1]) / (a - b) if abs(a - b) > 1e-2 else H_s[0][0] - H_s[0][1],
        (g_s[0] - g_s[2]) / (a - c) if abs(a - c) > 1e-2 else H_s[0][0] - H_s[0][2],
        (g_s[1] - g_s[2]) / (b - c) if abs(b - c) > 1e-2 else H_s[1][1] - H_s[1][2],
        S_s[0],
        S_s[1],
        S_s[2],
    ]

    Qmat = np.zeros((9, 9))
    for i, w in enumerate(Qs):
        Qmat[:, i] = mat2vec(U @ w @ Vt)
    H_rebuilt = Qmat @ np.diag(lmds) @ Qmat.T
    if verbose:
        print(f"--- ({m}) analytical hessian ---")
        print(H_rebuilt)

    ###############################################

    errors[m] = [
        LA.norm(g_F - g_F_rebuilt) / LA.norm(g_F),
        LA.norm(H_F - H_rebuilt) / LA.norm(H_F),
    ]

print("===== error summary =====")
for name, err in errors.items():
    print(f"{name}: grad: {err[0]:.3e}, Hess: {err[1]:.3e}")
